docs/app-config.rst
.. _app-config:

应用程序服务器全局设置
=========================

在阅读以下内容前，请确保您已经按照 :ref:`app-installation` 正确部署了应用程序服务器，并位于源码目录下（/path-to/zenith-cloud/app/）。

您可以根据您的环境修改源码目录下的 `config/config.go` 文件以配置服务器。

样例配置文件
----------------

您从 GitHub 仓库获取的源码中已经包含了一份默认的配置文件，忽略注释和函数部分后内容如下：

.. code-block:: go
   :linenos:
   
   const STORE_PATH = "G:\\Cloud\\"
   const DOWNLOAD_PATH = "G:\\Cloud\\"
   const CLIENT_VERSION = "Windows"
   const AUTHEN_BUFSIZE = 1024
   const BUFLEN = 4096 * 1024
   const MAXTRANSMITTER = 20
   const DATABASE_TYPE = "sqlite3"
   const DATABASE_PATH = "work.db"
   const START_USER_LIST = 10
   const TEST_USERNAME = "forec@bupt.edu.cn"
   const TEST_PASSWORD = "TESTTHISPASSWORD"
   const TEST_SAFELEVEL = 3
   const TEST_IP = "127.0.0.1"
   const TEST_PORT = 10087
   const SEPERATER = "+"
   const CHECK_LIVE_SEPERATE = 5
   const CHECK_LIVE_TAG = "[check]"
   const CHECK_MESSAGE_SEPERATE = 5
	
样例配置文件每项均有对应注释，您也可以查看下面的 :ref:`app-config-detailed` 了解每一项的具体功能。

.. _app-config-detailed:

样例文件详细解释
-------------------

配置文件中的每一项对应配置如下：

1. ``STORE_PATH`` ：服务器中存储所有用户上传文件的路径。

2. ``DOWNLOAD_PATH`` ：在 ``app/client`` 目录中提供了一个测试客户端，此记录用于设置测试客户端从服务器下载文件的存储路径。

3. ``CLIENT_VERSION`` ：测试用客户端使用的操作系统。

4. ``AUTHEN_BUFSIZE`` ：服务器和客户端之间命令交互连接使用的缓冲区大小，如果你不了解此字段的含义，可以保留默认值。此字段详细的介绍请参考 :ref:`app-protocal-authen-bufsize` 。

5. ``BUFLEN`` ：服务器和客户端之间传输长数据流时使用的缓冲区大小，在网络环境较差的情况下应降低该项的大小，在网络丢包率较低的情况下应增加该项的大小。如果你不了解此字段的含义，可以保留默认值。此字段的详细介绍请参考 :ref:`app-protocal-buflen` 。

6. ``MAXTRANSMITTER`` ：服务器为每个客户端允许分配的最大同时活动连接数量，增加此数值可以在用户尝试下载多个文件时提供并行能力。

7. ``DATABASE_TYPE`` ：服务器使用的数据库类型。

8. ``DATABASE_PATH`` ：服务器使用的数据库所在的路径，此路径可为绝对路径，也可为相对服务器可执行文件所在目录的相对路径。

9. ``START_USER_LIST`` ：服务器在启动后会维护一个活动的用户池，此项指定了服务器启动时分配的用户池大小，在加入新用户后无需分配新内存，除非用户池已满。

10. ``TEST_USERNAME`` ：客户端测试使用的用户名，你可以手动此用户加入数据库，或在服务器第一次启动时将此用户模型加入数据库。默认配置文件中的用户名、密码已经加入到仓库中的 ``work.db`` 数据库。

11. ``TEST_PASSWORD`` ：客户端测试使用的登陆密码。

12. ``TEST_SAFELEVEL`` ：测试客户端和服务器约定使用的安全等级，可设置为任意整数，小于等于 1 时被规整为 1，大于等于 3 时被规整为 3。1、2、3 等级分别对应在加密时使用 16、24、32字节的密钥。

13. ``TEST_IP`` ：测试服务器时开放的 IP 地址。

14. ``TEST_PORT`` ：测试服务器时开放的端口。

15. ``SEPERATER`` ：此项指定了服务器和客户端约定的命令格式使用的分隔符，默认情况下为 ``+`` 。如果你不了解命令格式，可以查看 :ref:`app-protocal-command` 。

16. ``CHECK_LIVE_SEPERATE`` ：此项指定了服务器保活检测的执行间隔，默认为 5。服务器会每隔 ``CHECK_LIVE_SEPERATE`` 秒向所有用户发送一条特定的消息以检查用户是否仍然存活，如果某个用户无法接收到服务器发送的消息，则认为该用户意外中断连接，服务器会登出该用户。

17. ``CHECK_LIVE_TAG`` ：此项指定了服务器保活检测时发送的特定消息前缀，默认为 ``[check]`` 。

18. ``CHECK_MESSAGE_SEPERATE`` ：此项指定了服务器转发消息执行的间隔，默认为 5。服务器会每隔 ``CHECK_MESSAGE_SEPERATE`` 秒向每个用户转发其他用户发送给他的消息。如果你不明白此项的意义，请查看 :ref:`app-protocal-chat` 。

例如，在使用 WiFi 连接的 Ubuntu 16.04 下部署服务器以供 Windows 客户端使用时，可参考的配置文件如下：

.. code-block:: go
   :linenos:
   
   const STORE_PATH = "/usr/local/cloud-store/"
   const DOWNLOAD_PATH = "/usr/local/cloud-download"
   const CLIENT_VERSION = "Windows"
   const AUTHEN_BUFSIZE = 1024
   const BUFLEN = 1024 * 1024  // 网络情况较差，减小缓冲区长度
   const MAXTRANSMITTER = 10
   const DATABASE_TYPE = "sqlite3"
   const DATABASE_PATH = "/usr/local/cloud/cloud.db"
   const START_USER_LIST = 10
   const TEST_USERNAME = "test@test.com"
   const TEST_PASSWORD = "TEST"
   const TEST_SAFELEVEL = 1
   const TEST_IP = "you ip address on Internet"	// 开放在公网地址
   const TEST_PORT = 10087
   const SEPERATER = "+"
   const CHECK_LIVE_SEPERATE = 10
   const CHECK_LIVE_TAG = "[check]"
   const CHECK_MESSAGE_SEPERATE = 10
   
接下来请您阅读 :ref:`app-quickstart` 。


docs/web-config.rst
.. _web-config:

Web 服务器全局配置
=====================

在阅读以下内容前，请确保您已经按照 :ref:`web-installation` 正确部署了顶点云 Web 服务器，并位于源码目录下（/path-to/zenith-cloud/web/）。

您可以根据您的环境修改源码目录下的 `config.py` 文件以配置服务器。

样例配置文件
----------------

您从 GitHub 仓库获取的源码中已经包含了一份默认的配置文件，这份配置文件中提供了一个基类 `Config` ，去掉注释后它的内容如下：

.. code-block:: python
   
   class Config:
    SECRET_KEY = os.environ.get('SECRET_KEY') or \
                 '9d0e91f3372224b3ec7afec2' \
                 '4313e745efcf00ba4a5b767b' \
                 '35b17834d5f26efac197fd69' \
                 'd881dd92e629dbfdc2f1fbf6'
    SQLALCHEMY_COMMIT_ON_TEARDOWN = True
    SQLALCHEMY_TRACK_MODIFICATIONS = True
    ZENITH_MAIL_SUBJECT_PREFIX = '[顶点云]'
    ZENITH_MAIL_SENDER = os.environ.get('ZENITH_MAIL_SENDER') or \
                         'cloud-storage@forec.cn'
    ZENITH_FILES_PER_PAGE = 10
    ZENITH_FOLLOWERS_PER_PAGE = 10
    ZENITH_COMMENTS_PER_PAGE = 10
    PROFILE_ZENITH_FILES_PER_PAGE = 6
    ZENITH_MESSAGES_PER_PAGE = 10
    ZENITH_TEMPFOLDER_LENGTH = 12
    ZENITH_PATH_SEPERATOR = '\\'
    ZENITH_FILE_STORE_PATH = 'G:\\Cloud\\'
    ZENITH_TEMPFILE_STORE_PATH = ZENITH_FILE_STORE_PATH + \
                             'TEMP' + ZENITH_PATH_SEPERATOR
    ZENITH_FOLDER_ZIP_SUFFIX = 'zenith'
    ZENITH_INVALID_INFFIX = ['//', '\\', '/', '..', '%', '^', '&',
                             '*', '$', '!', '+', '#']
    EMAIL_ADMIN ='forec@bupt.edu.cn'
    ZENITH_RANDOM_PATH_ELEMENTS = ['a', 'b', 'c', 'd', 'e', 'f', 'g',
                                   'h', 'i', 'j', 'k', 'l', 'm', 'n',
                                   'o', 'p', 'q', 'r', 's', 't', 'u',
                                   'v', 'w', 'x', 'y', 'z', '1', '2',
                                   '3', '4', '5', '6', '7', '8', '9',
                                   '0', 'A', 'B', 'C', 'D', 'E', 'F',
                                   'G', 'H', 'I', 'J', 'K', 'L', 'M',
                                   'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                                   'U', 'V', 'W', 'X', 'Y', 'Z']
    ZENITH_VALID_THUMBNAIL = ['.jpg', '.png', '.ico', '.jpeg']
    ZENITH_VALID_THUMBNAIL_SIZE = 512 * 1024

	
样例配置文件中，每项均有对应注释，您也可以查看下面的 :ref:`web-config-detailed` 了解每一项的具体功能。

.. _web-config-detailed:

样例文件详细解释
-------------------

配置文件中 `Config` 类的每一项对应配置如下：

1. ``SECRET_KEY`` ：用于服务器生成 token 使用的密钥，不可泄露，应当设置在服务器部署系统的环境变量中
2. ``SQLALCHEMY_COMMIT_ON_TEARDOWN`` ：为 ``True`` 时服务器终止运行时向数据库提交变动
3. ``SQLALCHEMY_TRACK_MODIFICATIONS`` ：为 ``True`` 时数据库将追踪服务器对数据的改动
4. ``ZENITH_MAIL_SUBJECT_PREFIX`` ：服务器向用户发送的验证邮件的主题前缀
5. ``ZENITH_MAIL_SENDER`` ：服务器向用户发送验证邮件使用的邮箱
6. ``ZENITH_FILES_PER_PAGE`` ：Index 页面每页显示的文件数量
7. ``ZENITH_FOLLOWERS_PER_PAGE`` ：关注着界面每页显示的用户数量
8. ``ZENITH_COMMENTS_PER_PAGE`` ：文件详细页面每页显示的评论数量
9. ``PROFILE_ZENITH_FILES_PER_PAGE`` ：用户资料页每页显示的文件数量
10. ``ZENITH_MESSAGES_PER_PAGE`` ：用户消息页面每页显示的消息数量
11. ``ZENITH_TEMPFOLDER_LENGTH`` ：服务器在临时存储文件时生成的随机目录名长度
12. ``ZENITH_PATH_SEPERATOR`` ：服务器所属文件系统的目录分隔符，Windows为 *\* ，Linux 为 */*
13. ``ZENITH_FILE_STORE_PATH`` ：服务器用于存储用户文件的路径
14. ``ZENITH_TEMPFILE_STORE_PATH`` ：服务器生成的随机目录的根路径，默认为文件存储路径下的 TMEP 文件夹
15. ``ZENITH_FOLDER_ZIP_SUFFIX`` ：用户上传目录压缩包时使用的后缀，使用此后缀的文件会被视作一个目录
16. ``ZENITH_INVALID_INFFIX`` ：此列表中的字符不能出现在用户文件名中，否则视为不合法
17. ``EMAIL_ADMIN`` ：管理员账户使用的邮箱
18. ``ZENITH_RANDOM_PATH_ELEMENTS`` ：服务器生成的随机路径包含的元素
19. ``ZENITH_VALID_THUMBNAIL`` ：服务器允许用户上传的头像后缀名
20. ``ZENITH_VALID_THUMBNAIL_SIZE`` ：服务器允许用户上传的最大头像大小

`Config` 类仅仅是配置类的基类，你需要扩展此类才可完成配置。在 `config.py` 文件中，存在一些默认的 `Config` 子类，如 `LinuxConfig` 、 `WindowsConfig` 。

下面简单介绍 `LinuxConfig` 并以一个环境为例讲解如何配置。

`LinuxConfig` 类的内容如下，它是我（ `Forec`_ ）在我的云主机部署 Web 服务器时使用的配置类：

.. code-block:: python
   
   class LinuxConfig(Config):
    ZENITH_PATH_SEPERATOR = '/'    # 服务器所属文件系统的目录分隔符
    ZENITH_FILE_STORE_PATH = '/root/work/cloud/Cloud/'  # 服务器存储用户文件的路径
    ZENITH_TEMPFILE_STORE_PATH = ZENITH_FILE_STORE_PATH + \
                                 'TEMP' + ZENITH_PATH_SEPERATOR
    ZENITH_SERVER_ADDRESS = 'cloud.forec.cn' # 服务器部署的域名/IP地址
    SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD')

可以看出， `LinuxConfig` 类重写了 `Config` 类的几项，同时添加了几个新的选项。新选项介绍如下：

1. ``ZENITH_SERVER_ADDRESS`` ：服务器部署使用的域名/IP地址
2. ``SERVER_NAME`` ：Flask 中的 *url_for* 函数使用的服务器名，通常保持和 *SERVER_NAME* 一致
3. ``SQLALCHEMY_DATABASE_URI`` ：服务器使用的数据库所在的路径
4. ``MAIL_SERVER`` ：服务器发送邮件使用的邮箱服务器
5. ``MAIL_PORT`` ：服务器使用 stmp 协议的端口号，通常为 25。使用 SSL 时设置为 465 但这取决于 *MAIL_SERVER* 是否支持 SSL
6. ``MAIL_USE_TLS`` ：是否启用安全连接发送邮件，通常设置为 ``True`` 
7. ``MAIL_USERNAME`` ：服务器发送邮件使用的邮箱帐号，通常和 *ZENITH_MAIL_SENDER* 保持一致
8. ``MAIL_PASSWORD`` ：服务器发送邮件使用的邮箱帐号的密码，通常保存在环境变量中

.. _web-config-example:

下面通过一个实例环境解释如何配置。

例如，在安装 Ubuntu 16.04 的主机上部署 Web 服务器，可参考的配置文件如下（使用扩展类）：

.. code-block:: python
   
   class MyConfig(Config):
    ZENITH_PATH_SEPERATOR = '/'
    ZENITH_FILE_STORE_PATH = '/home/cloud/'  # 将用户上传文件存储在 /home/cloud 中
    ZENITH_TEMPFILE_STORE_PATH = ZENITH_FILE_STORE_PATH + \
                                 'TEMP' + ZENITH_PATH_SEPERATOR
    ZENITH_SERVER_ADDRESS = 'myaddress.my.io' # 自定义的域名，你需要先购买此域名并且映射到部署主机上
    SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = '/usr/local/cloud/mydb.db'
    	# 设置数据库为 /usr/local/cloud/mydb.db
    MAIL_SERVER = 'smtp.163.com'	# 使用 163 邮箱
    MAIL_PORT = 25
    MAIL_USE_TLS = True
    MAIL_USERNAME = "mycloud@163.com"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or '123456'
    	# 在环境变量中添加密码，若环境变量未找到对应值则使用 123456
    	
.. _web-config-add:

添加自定义配置类到表
------------------------

在已经定义了自定义配置类后，你需要将自定义配置类添加到表驱动中以使 :ref:`web-app-factory` 能根据我的配置类生成服务器实例。

在 `config.py` 中，有一个名为 `config` 的字典如下：

.. code-block:: python

   config = {
    'development' : DevelopmentConfig,      # 开发环境
    'linux': LinuxConfig,                   # 提供的 Linux 模板环境
    'windows': WindowsConfig,               # 提供的 Windows模板环境
    'testing' : TestingConfig,              # 测试环境
    'default' : DevelopmentConfig           # 默认为开发环境
    }
    
你需要添加自己的自定义配置类到此表中，如添加 ``'myconfig': MyConfig`` 。之后，修改 `manage.py` 中的第 13 行 ``app = create_app('default')`` 为 ``app = create_app('myconfig')`` 即可。
   
接下来请您阅读 :ref:`web-quickstart` 。

.. _Forec: http://forec.cn


web/config.py
# 作者：Forec
# 最后修改时间：2016-12-9
# 邮箱：forec@bupt.edu.cn
# 关于此文件: 服务器所有的配置信息在此文件中指定，具体功能由功能项后
#    的注释标明

import os
basedir = os.path.abspath(os.path.dirname(__file__))

class Config:
    SECRET_KEY = os.environ.get('SECRET_KEY') or \
                 '9d0e91f3372224b3ec7afec2' \
                 '4313e745efcf00ba4a5b767b' \
                 '35b17834d5f26efac197fd69' \
                 'd881dd92e629dbfdc2f1fbf6'         # 用于为安全操作生成 token 的密钥，不可泄露
    SQLALCHEMY_COMMIT_ON_TEARDOWN = True            # 数据库在服务器终止时 commit 变动
    SQLALCHEMY_TRACK_MODIFICATIONS = True           # 数据库追踪改动
    ZENITH_MAIL_SUBJECT_PREFIX = '[顶点云]'         # 服务器发送验证邮件的主题前缀
    ZENITH_MAIL_SENDER = os.environ.get('ZENITH_MAIL_SENDER') or \
                         'cloud-storage@forec.cn'   # 服务器向用户发送验证邮件的邮箱
    ZENITH_FILES_PER_PAGE = 10              # Index 页面每页显示的文件数量
    ZENITH_FOLLOWERS_PER_PAGE = 10          # 每页显示的关注者数量
    ZENITH_COMMENTS_PER_PAGE = 10           # 每页显示的评论数量
    PROFILE_ZENITH_FILES_PER_PAGE = 6       # 用户资料页每页显示的文件数量
    ZENITH_MESSAGES_PER_PAGE = 10           # Message 页面每页显示的消息数量
    ZENITH_TEMPFOLDER_LENGTH = 12   # 服务器生成的随机目录名长度
    ZENITH_PATH_SEPERATOR = '\\'    # 服务器所属文件系统的目录分隔符，Windows为\\，*nix 为//
    ZENITH_FILE_STORE_PATH = 'G:\\Cloud\\'  # 服务器存储用户文件的路径
    ZENITH_TEMPFILE_STORE_PATH = ZENITH_FILE_STORE_PATH + \
                                 'TEMP' + ZENITH_PATH_SEPERATOR
        # 服务器生成随机目录所在的路径，默认为文件存储路径下的 TMEP 文件夹
    ZENITH_FOLDER_ZIP_SUFFIX = 'zenith'
    ZENITH_INVALID_INFFIX = ['//', '\\', '/', '..', '%', '^', '&',
                             '*', '$', '!', '+', '#']
    EMAIL_ADMIN ='forec@bupt.edu.cn'                # 管理员账户的邮箱
    ZENITH_RANDOM_PATH_ELEMENTS = ['a', 'b', 'c', 'd', 'e', 'f', 'g',
                                       'h', 'i', 'j', 'k', 'l', 'm', 'n',
                                       'o', 'p', 'q', 'r', 's', 't', 'u',
                                       'v', 'w', 'x', 'y', 'z', '1', '2',
                                       '3', '4', '5', '6', '7', '8', '9',
                                       '0', 'A', 'B', 'C', 'D', 'E', 'F',
                                       'G', 'H', 'I', 'J', 'K', 'L', 'M',
                                       'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                                       'U', 'V', 'W', 'X', 'Y', 'Z']
        # 用于生成随机路径的元素
    ZENITH_VALID_THUMBNAIL = ['.jpg', '.png', '.ico', '.jpeg']
        # 允许用户使用的头像后缀名
    ZENITH_VALID_THUMBNAIL_SIZE = 512 * 1024 # 最大 512K 头像

    @staticmethod
    def init_app(app):
        pass

class DevelopmentConfig(Config):            # 开发者环境配置
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or \
                    "Cloud-Storage-2016"

class TestingConfig(Config):                # 测试环境配置
    TESTING = True
    WTF_CSRF_ENABLED = False
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'temp/test.sqlite')
    ZENITH_PATH_SEPERATOR = '/'
    ZENITH_FILE_STORE_PATH = 'temp/store/'
    ZENITH_TEMPFILE_STORE_PATH = ZENITH_FILE_STORE_PATH + \
                                 'TEMP' + ZENITH_PATH_SEPERATOR
    ZENITH_SERVER_ADDRESS = '127.0.0.1'

class WindowsConfig(Config):
    ZENITH_PATH_SEPERATOR = '\\'    # 服务器所属文件系统的目录分隔符，Windows为\\，*nix 为//
    ZENITH_FILE_STORE_PATH = 'G:\\Cloud\\'  # 服务器存储用户文件的路径
    ZENITH_TEMPFILE_STORE_PATH = ZENITH_FILE_STORE_PATH + \
                                 'TEMP' + ZENITH_PATH_SEPERATOR
    #ZENITH_SERVER_ADDRESS = '127.0.0.1'# or 'cloud.forec.cn' # 服务器部署的域名/IP地址
    #SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or \
                    "Cloud-Storage-2016"

class LinuxConfig(Config):
    ZENITH_PATH_SEPERATOR = '/'    # 服务器所属文件系统的目录分隔符，Windows为\\，*nix 为//
    ZENITH_FILE_STORE_PATH = '/root/work/cloud/Cloud/'  # 服务器存储用户文件的路径
    ZENITH_TEMPFILE_STORE_PATH = ZENITH_FILE_STORE_PATH + \
                                 'TEMP' + ZENITH_PATH_SEPERATOR
    ZENITH_SERVER_ADDRESS = 'cloud.forec.cn' # 服务器部署的域名/IP地址
    SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or \
                    "Cloud-Storage-2016"

config = {
    'development' : DevelopmentConfig,      # 开发环境
    'linux': LinuxConfig,                   # 提供的 Linux 模板环境
    'windows': WindowsConfig,               # 提供的 Windows模板环境
    'testing' : TestingConfig,              # 测试环境
    'default' : TestingConfig               # 默认为测试环境
}
